cpp/tests/test_metadata_payload_exchange.cpp

```diff
@@ -0,0 +1,314 @@
+        auto messages = comm_interface->recv();
+        std::ranges::move(messages, std::back_inserter(received_messages));
+
+        if (received_messages.empty()) {
+            std::this_thread::yield();
+        }
+    }
+
+    EXPECT_EQ(received_messages.size(), 1);
+    auto& msg = received_messages[0];
+    EXPECT_EQ(msg->peer_rank(), prev_rank);
+    EXPECT_EQ(msg->metadata(), test_metadata);
+    verify_data_content(msg->data(), data_size);
+
+    wait_for_communication_complete();
+}
+
+TEST_F(MetadataPayloadExchangeTest, MultipleMessages) {
+    if (comm->nranks() < 2) {
+        GTEST_SKIP() << "Test requires at least 2 ranks";
+    }
+
+    Rank next_rank = (comm->rank() + 1) % comm->nranks();
+    Rank prev_rank = (comm->rank() - 1 + comm->nranks()) % comm->nranks();
+    constexpr int num_messages = 5;
+
+    // All ranks send multiple messages to next rank
+    std::vector<std::unique_ptr<MetadataPayloadExchange::Message>> messages;
+
+    for (int i = 0; i < num_messages; ++i) {
+        std::vector<std::uint8_t> metadata = {
+            static_cast<std::uint8_t>(i),
+            static_cast<std::uint8_t>(i + 1),
+            static_cast<std::uint8_t>(i + 2)
+        };
+        std::size_t data_size =
+            (i % 2 == 0)
+                ? 0
+                : (i + 1) * 100;  // Alternate between metadata-only and with-data
+        messages.push_back(create_test_message(next_rank, metadata, data_size));
+    }
+
+    comm_interface->send(std::move(messages));
+
+    // All ranks receive from previous rank
+    std::vector<std::unique_ptr<MetadataPayloadExchange::Message>> received_messages;
+    while (received_messages.size() < num_messages) {
+        comm_interface->progress();
+        auto messages = comm_interface->recv();
+        std::ranges::move(messages, std::back_inserter(received_messages));
+
+        if (received_messages.empty()) {
+            std::this_thread::yield();
+        }
+    }
+
+    // All ranks should get all messages from previous rank
+    EXPECT_EQ(received_messages.size(), num_messages);
+
+    for (int i = 0; i < static_cast<int>(received_messages.size()); ++i) {
+        auto& msg = received_messages[i];
+        EXPECT_EQ(msg->peer_rank(), prev_rank);
+
+        // Check metadata
+        std::vector<std::uint8_t> expected_metadata = {
+            static_cast<std::uint8_t>(i),
+            static_cast<std::uint8_t>(i + 1),
+            static_cast<std::uint8_t>(i + 2)
+        };
+        EXPECT_EQ(msg->metadata(), expected_metadata);
+
+        // Check data presence and content
+        if (i % 2 == 0) {
+            EXPECT_EQ(msg->data(), nullptr);  // Even indices are metadata-only
+        } else {
+            EXPECT_NE(msg->data(), nullptr);  // Odd indices have data
+            std::size_t expected_size = (i + 1) * 100;
+            verify_data_content(msg->data(), expected_size);  // Odd indices have data
+        }
+    }
+
+    wait_for_communication_complete();
+}
```

Comment:
Can you add a test that demonstrates "half" a ping-pong. i.e. `rank-0` sends to `rank-1`, `rank-1` receives from `rank-0`. But `rank-1` never sends anything, and `rank-0` never receives anything.

Reply:
But how do you then identify the test is "done"? A timeout? Or are you suggesting we implement support for canceling requests in the `Communicator`? We currently don't implement that for either MPI or UCXX. I don't know if MPI cancelation is deterministic, in UCXX it depends on the transports that are actually enabled and requires that we enable error handling which we currently do not, thus I don't think we can deterministically verify this.

Reply:
The test is done when rank 0 has finished sending and rank 1 has finished receiving. If this is not a pattern we can support then we need to rethink